verl/workers/actor/dp_actor.py

```diff
@@ -62,14 +64,29 @@ def _forward_micro_batch(self, micro_batch, temperature) -> Tuple[torch.Tensor,
                 # unpad the position_ids to align the rotary
                 position_ids_rmpad = index_first_axis(rearrange(position_ids.unsqueeze(-1), "b s ... -> (b s) ..."),
                                                       indices).transpose(0, 1)
+
+                # pad and slice the inputs if sp > 1
+                origin_input_ids_rmpad = input_ids_rmpad
+                if self.ulysses_sequence_parallel_size > 1:
+                    input_ids_rmpad, position_ids_rmpad, pad_size = ulysses_pad_and_slice_inputs(input_ids_rmpad, \
+                                                                                                position_ids_rmpad, \
+                                                                                                sp_size=self.ulysses_sequence_parallel_size)
                 # only pass input_ids and position_ids to enable flash_attn_varlen
                 output = self.actor_module(input_ids=input_ids_rmpad,
                                            attention_mask=None,
                                            position_ids=position_ids_rmpad,
                                            use_cache=False)  # prevent model thinks we are generating
                 logits_rmpad = output.logits.squeeze(0)  # (total_nnz, vocab_size)
+
+                # gather output if sp > 1
+                if self.ulysses_sequence_parallel_size > 1:
+                    logits_rmpad = gather_outpus_and_unpad(logits_rmpad,
```

Comment:
It's problematic to perform any operation on full logits. We should perform reduction on logits first, then gather